packages/flutter/lib/src/gestures/selection_recognizers.dart

```diff
@@ -0,0 +1,701 @@
+typedef GestureTapUpWithTapStatusCallback  = void Function(TapUpDetails details, TapStatus status);
+
+/// {@macro flutter.gestures.dragdetails.GestureDragStartCallback}
+///
+/// The consecutive tap count when the drag was initiated is given by [TapStatus.consecutiveTapCount].
+///
+/// Used by [TapAndDragGestureRecognizer.onStart].
+typedef GestureDragStartWithTapStatusCallback = void Function(DragStartDetails details, TapStatus status);
+
+/// {@macro flutter.gestures.dragdetails.GestureDragUpdateCallback}
+///
+/// The consecutive tap count when the drag was initiated is given by [TapStatus.consecutiveTapCount].
+///
+/// Used by [TapAndDragGestureRecognizer.onUpdate].
+typedef GestureDragUpdateWithTapStatusCallback = void Function(DragUpdateDetails details, TapStatus status);
+
+/// {@macro flutter.gestures.monodrag.GestureDragEndCallback}
+///
+/// The consecutive tap count when the drag was initiated is given by [TapStatus.consecutiveTapCount].
+///
+/// Used by [TapAndDragGestureRecognizer.onEnd].
+typedef GestureDragEndWithTapStatusCallback = void Function(DragEndDetails endDetails, TapStatus status);
+
+mixin _ConsecutiveTapMixin {
+  // For consecutive tap
+  Timer? consecutiveTapTimer;
+  Offset? lastTapOffset;
+  int consecutiveTapCount = 0;
+
+  bool isWithinConsecutiveTapTolerance(Offset secondTapOffset) {
+    assert(secondTapOffset != null);
+    if (lastTapOffset == null) {
+      return false;
+    }
+
+    final Offset difference = secondTapOffset - lastTapOffset!;
+    return difference.distance <= kDoubleTapSlop;
+  }
+
+  void incrementConsecutiveTapCountOnDown(Offset tapGlobalPosition) {
+    if (lastTapOffset == null) {
+      // If last tap offset is null then we have not started our consecutive tap count,
+      // so the consecutiveTapTimer should be null.
+      assert(consecutiveTapTimer == null);
+      consecutiveTapCount += 1;
+      lastTapOffset = tapGlobalPosition;
+    } else if (consecutiveTapTimer != null && isWithinConsecutiveTapTolerance(tapGlobalPosition)) {
+      consecutiveTapCount += 1;
+      consecutiveTapTimerStop();
+    }
+  }
+
+  void consecutiveTapReset() {
+    consecutiveTapTimer?.cancel();
+    consecutiveTapTimer = null;
+    lastTapOffset = null;
+    consecutiveTapCount = 0;
+  }
+
+  void consecutiveTapTimerStop() {
+    consecutiveTapTimer?.cancel();
+    consecutiveTapTimer = null;
+  }
+}
+
+/// An object that includes supplementary details of a tap event, such as
+/// if the shift key was pressed when the tap occured, and what the tap count
+/// is.
+class TapStatus {
+  /// Creates a [TapStatus].
+  const TapStatus({
+    required this.consecutiveTapCount,
+    required this.isShiftPressed,
+  });
+
+  /// If this tap is in a series of taps, the `consecutiveTapCount` is
+  /// what number in the series this tap is.
+  final int consecutiveTapCount;
+
+  /// Whether the shift key was pressed when this tap happened.
+  final bool isShiftPressed;
```

Comment:
Nit: Maybe `shiftWasDown` or something like that?  I don't know if I like that any better...  But it's slightly confusing that this means that the shift was down at the time of the tap, not necessarily when it is accessed.

Reply:
I changed this to `keysPressedOnDown` in case there are other keyboard + gesture combinations that need to be handled in the future. I have so far identified one other gesture besides shift, which is ctrl. ctrl + click acts like a right click on macOS, selecting the word at the clicked position and popping up the context menu.

Reply:
Ah good call!